File: Src/ButterflyNet.Parser/Strategies/HeaderStrategy.cs
using ButterflyNet.Parser.Satisfiers;
using ButterflyNet.Parser.Scopes;

namespace ButterflyNet.Parser.Strategies {
	public class HeaderStrategy : ScopeDrivenStrategy {
		public HeaderStrategy() {
			AddSatisfier<StartOfLineSatisfier>();
			AddSatisfier(new ExactCharMatchSatisfier("!"));
		}

		protected override void DoExecute(ParseContext context) {
			var depth = 1;
			while (context.Input.Peek() == '!') {
				depth++;
				context.Input.Read();
			}

			context.Input.SeekToNonWhitespace(); //ignore spaces/tabs
			context.UpdateCurrentChar();
			OpenScope(new HeaderScope(depth), context);
		}
	}
}

File: src/js/src/strategies/HeaderStrategy.js
function HeaderStrategy() {
	HeaderStrategy.$parent.call(this);
	
	this.addSatisfier(new StartOfLineSatisfier());
	this.addSatisfier(new CannotNestInsideInlineSatisfier());
	
	this.setAsTokenTransformer("!");
	
	this.doExecute = function(context) {
		var depth = 1;
		while (context.input.peek() === "!") {
			depth++;
			context.input.read();
		}
		
		context.input.seekToNonWhitespace();
		context.updateCurrentChar();
		this.openScope(new HeaderScope(depth), context);
	};
}

extend(ScopeDrivenStrategy, HeaderStrategy);